libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -5960,6 +5960,51 @@ void GCS_MAVLINK::send_autopilot_state_for_gimbal_device() const
 #endif  // AP_AHRS_ENABLED
 }
 
+#if AP_MAVLINK_MSG_FLIGHT_INFORMATION_ENABLED
+void GCS_MAVLINK::send_flight_information()
+{
+    const uint32_t time_boot_ms = AP_HAL::millis();
+
+    // We send time since boot (in micros) rather than a UTC timestamp, as this
+    // works better with SITL when SIM_SPEEDUP > 1
+    const uint64_t arm_time_us = (uint64_t)AP::arming().arm_time_ms() * AP_USEC_PER_MSEC;
+
+    const MAV_LANDED_STATE current_landed_state = landed_state();
+    if (flight_info.last_landed_state != current_landed_state) {
+        switch (current_landed_state) {
+            case MAV_LANDED_STATE_IN_AIR:
+            case MAV_LANDED_STATE_TAKEOFF:
+            case MAV_LANDED_STATE_LANDING: {
+                if (!flight_info.takeoff_time_ms) {
+                    flight_info.takeoff_time_ms = time_boot_ms;
+                }
+                break;
+            }
+            default: {
+                flight_info.takeoff_time_ms = 0;
+                break;
+            }
+        }
+
+        flight_info.last_landed_state = current_landed_state;
+    }
+
+    // We send time since boot (in micros) rather than a UTC timestamp, as this
+    // works better with SITL when SIM_SPEEDUP > 1
```

Comment:
This comment is out of date once the patches to the mavlink message upstream are brought back.

Instead it should read something like the fields are misnamed...

Also, the flag value is a little problematic with the problematic `ARMING_REQUIRE` flag on Rover and Plane.  There's always an argument that they're not really armed in that state, but just pretend to be.... but still, how does this message work for an `ARMING_REQUIRE` 0 vehicle?

Reply:
> [...] There's always an argument that they're not really armed in that state, but just pretend to be...

They will set the `MAV_MODE_FLAG_SAFETY_ARMED` flag of the `base_mode` field of the `HEARTBEAT` message, so as far as a GCS knows, they _are_ armed.

> [...] how does this message work for an `ARMING_REQUIRE` 0 vehicle?

It will report arming time as `0` if `AP_Arming::arm()` is never called. This field shouldn't be treated as a proxy for "is armed", it needs to be read in conjunction with the arming sate reported in the `HEARTBEAT`.